### ISSUE TITLE
Normalized enrichment score = negative infinity and results not reproducible across different computers

### ISSUE THREAD

**Issue by robinpaul85** [MaayanLab/blitzgsea#36]
Many times I see the normalized enrichment score to be infinity or negative infinity. Also in general, i find the results to be not reproducible when I run the same input in different computers. I see the code base makes extensive use of random numbers. Maybe set a seed for all random numbers?

<img width="555" alt="Screenshot 2024-10-04 at 1 35 46 PM" src="https://github.com/user-attachments/assets/f3263d58-ad86-46ad-bdd2-a6960e342122">

**Comment by lachmann12** [comment]
Hi, thank you for the feedback. Can you confirm that this is happening on the latest version of blitzgsea? The latest version is 1.3.47 which did address the inf issue. If this is still something you experience is it possible to share the signature that causes this so I can try to reproduce the error?

**Comment by robinpaul85** [comment]
Is there a way to check the blitzgsea version ?

**Comment by lachmann12** [comment]
You can use the following command line call:

pip3 freeze | grep blitzgsea

**Comment by lachmann12** [comment]
if you are using windows:

pip3 freeze | findstr blitzgsea

**Comment by robinpaul85** [comment]
Yes, I did upgrade the blitzgsea version. It still shows NES=-inf. The most difficult part is its hard to reproduce the results even on the same computer. Sometimes it shows a non-infinite number and sometimes infinity.

**Comment by stevendbrown** [comment]
I resolved an apparently identical issue by increasing the `deep_accuracy` argument to `.gsea()` from 50 (default on my install) to 200, and increasing the `accuracy` argument from 40 to 100.

**Comment by stevendbrown** [comment]
I ran into another place where NES=-inf despite increases to `accuracy` & `deep_accuracy`. The primary problem seemed to be that `pos_ratio` would come out to 1, which lead to non-finite values in the NES calculation roughly [here](https://github.com/MaayanLab/blitzgsea/blob/b935dbb065d433d9bb0864b9327da6dc414214e8/blitzgsea/__init__.py#L336). In those cases, increasing the `permutations` argument from 1000 to 2000 helped push them to finite values most of the time, varying with random seed provided to `.gsea()`. Increasing the permutations even more produced more reliably finite NES values, but the NES and resulting p values had marked variation depending on the seed.

While trying to hunt down the variation, I experimented with changing the scale value in [this line](https://github.com/MaayanLab/blitzgsea/blob/b935dbb065d433d9bb0864b9327da6dc414214e8/blitzgsea/__init__.py#L154). Changing the scale from 1e-4 to 1e-3 further reduced the variation in NES values. Combining this increase of the scale at the linked line with the increased permutations resulted in more stable NES values, but some seeds still have non-finite NES or NES with fairly large deviations from other seeds.

Altering the number of anchors also seems to significantly affect the results. There's enough moving parts here that I feel like I'm missing something important that may underpin the variations.

**Comment by robinpaul85** [comment]
In the results, for some pathways there is a finite NES value (and p-values are significant) but no leading edge genes. Is this expected ?

**Comment by lachmann12** [comment]
It seems not possible. It would be great if I could somehow replicate your issue. If there is no leading edge I suspect that the gene identifiers in the gene set library do not match the identifiers in the signature. How large is your gene expression signature?

**Comment by lachmann12** [comment]
@stevendbrown if the p-values are too small there might be numeric instability. There have been some fixes that should improve the numeric stability, but they are a couple of months ago. Make sure the latest version is installed. Since the null model is computed with randomized running sums the p-values can vary slightly. But for sufficiently large number of randomizations this should be fairly stable.

**Comment by lachmann12** [comment]
I believe to have fixed an issue that caused some problems with the NES calculation, which stems from the estimation of the gamma distribution parameters. In my tests this solved the issues I could come up with. The new version on pypi or github should work better now.
